File: src/main/java/com/github/solairerove/algs4/leprosorium/arrays/MonotonicArray.java
package com.github.solairerove.algs4.leprosorium.arrays;

import java.util.Arrays;

/**
 * from left to right non increase or non decrease
 */
public class MonotonicArray {
    public static void main(String[] args) {
        int[] a = {-2, -5, -5, -6, -20, -30, -1000};
        System.out.printf("%s %s", Arrays.toString(a), isMonotonic(a));
    }

    private static boolean isMonotonic(int[] a) {
        int n = a.length;

        boolean inc = true;
        boolean dec = true;

        for (int i = 1; i < n; i++) {
            if (a[i] > a[i - 1]) {
                inc = false;
            }

            if (a[i] < a[i - 1]) {
                dec = false;
            }
        }

        return inc || dec;
    }
}

File: src/main/kotlin/com/github/solairerove/algs4/leprosorium/arrays/MonotonicArray.kt
package com.github.solairerove.algs4.leprosorium.arrays

fun main() {
    val arr = listOf(-10, -9, -4, 0, 1, 2, 3, 5, 9, 11, 12, 14, 17)
    print(isMonotonic(arr))
}

// O(n) time | O(1) space
private fun isMonotonic(arr: List<Int>): Boolean {
    var isIncreasing = true
    var isDecreasing = true

    for (i in 1 until arr.size) {
        when {
            arr[i - 1] < arr[i] -> isDecreasing = false
            arr[i - 1] > arr[i] -> isIncreasing = false
        }
    }

    return isIncreasing || isDecreasing
}
